soc/nordic/common/power.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/poweroff.h>
+#include <zephyr/toolchain.h>
+#include <zephyr/pm/policy.h>
+#include <zephyr/arch/common/pm_s2ram.h>
+#include <hal/nrf_resetinfo.h>
+#include <hal/nrf_lrcconf.h>
+#include <hal/nrf_memconf.h>
+#include <zephyr/cache.h>
+#include <power.h>
+#include <pm_s2ram.h>
+
+static void suspend_common(void)
+{
+
+	/* Flush, disable and power down DCACHE */
+	sys_cache_data_flush_all();
+	sys_cache_data_disable();
+	nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+					RAMBLOCK_CONTROL_BIT_DCACHE, false);
+
+	if (IS_ENABLED(CONFIG_ICACHE)) {
+		/* Flush, disable and power down ICACHE */
+		sys_cache_instr_disable();
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+						RAMBLOCK_CONTROL_BIT_ICACHE, false);
+	}
+
+	/* Disable retention */
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+}
+
+void nrf_poweroff(void)
+{
+	nrf_resetinfo_resetreas_local_set(NRF_RESETINFO, 0);
+	nrf_resetinfo_restore_valid_set(NRF_RESETINFO, false);
+
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+
+	/* TODO: Move it around k_cpu_idle() implementation. */
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+
+	suspend_common();
+
+	nrf_lrcconf_task_trigger(NRF_LRCCONF010, NRF_LRCCONF_TASK_SYSTEMOFFREADY);
+
+	__set_BASEPRI(0);
```

Comment:
It should rather be `cpu_idle()`. It contains additional features (pre and post hooks, etc.)

Reply:
@hubertmis , what do you think? As far as I remember you wanted me to replace `k_cpu_idle()` here this way.

Reply:
Absolutely! Exactly because of the hooks which cannot be executed in S2RAM procedure. S2RAM requires other set of hooks (if any) than S2Idle